tests/h/services/oauth_test.py

```diff
@@ -209,6 +209,66 @@ def test_it_raises_when_client_authority_does_not_match_userid(self, svc, db_ses
         assert exc.value.type == 'invalid_grant'
         assert 'authenticated client and JWT subject authorities do not match' in exc.value
 
+
+    @pytest.mark.parametrize('grant_start,grant_expiry',
+                             [[None, timedelta(minutes=15)],
+                              [timedelta(minutes=-15), None],
+                              [timedelta(minutes=-9), timedelta(minutes=9)]])
+    def test_overlong_expiry(self, svc, claims, authclient, jwt_bearer_body, grant_start, grant_expiry):
+        claims['nbf'] = self.epoch(delta=grant_start)
+        claims['exp'] = self.epoch(delta=grant_expiry)
+        jwt_bearer_body['assertion'] = self.jwt_token(claims, authclient.secret)
+
+        with pytest.raises(OAuthTokenError) as exc:
+            svc.verify_token_request(jwt_bearer_body)
+
+        assert exc.value.type == 'invalid_grant'
+
+    def test_missing_expiry(self, svc, claims, authclient, jwt_bearer_body):
+        del claims['exp']
+        jwt_bearer_body['assertion'] = self.jwt_token(claims, authclient.secret)
+
+        with pytest.raises(OAuthTokenError) as exc:
+            svc.verify_token_request(jwt_bearer_body)
+
+        assert exc.value.type == 'invalid_grant'
+        assert 'JWT is missing claim exp' in exc.value
+
+    def test_missing_nbf(self, svc, claims, authclient, jwt_bearer_body):
+        del claims['nbf']
+        jwt_bearer_body['assertion'] = self.jwt_token(claims, authclient.secret)
+
+        with pytest.raises(OAuthTokenError) as exc:
+            svc.verify_token_request(jwt_bearer_body)
+
+        assert exc.value.type == 'invalid_grant'
+        assert 'JWT is missing claim nbf' in exc.value
+
+    @pytest.mark.parametrize('claim_name', ['nbf', 'exp'])
+    def test_null_timestamp(self, svc, claims, authclient, jwt_bearer_body, claim_name):
+        claims[claim_name] = None
+        jwt_bearer_body['assertion'] = self.jwt_token(claims, authclient.secret)
+
+        with pytest.raises(OAuthTokenError) as exc:
+            svc.verify_token_request(jwt_bearer_body)
+
+        assert exc.value.type == 'invalid_grant'
+        assert 'JWT is missing claim {}'.format(claim_name) in exc.value
+
+    @pytest.mark.parametrize('claim_name,delta',
+                             [['iat', timedelta(minutes=-5)],
+                              ['nbf', timedelta(minutes=-5)],
+                              ['exp', timedelta(minutes=5)]])
+    def test_string_timestamp(self, svc, claims, authclient, jwt_bearer_body, claim_name, delta):
+        claims[claim_name] = unicode(self.epoch(delta=delta))
```

Comment:
Ah, yes – I’d seen that dotted around the codebase before, but I hadn’t made the connection that was what it was for. Thanks for the tip. 👍